### PR TITLE
Provide default Canvas API scopes for OAuth, with an environment variable escape hatch (#202)

### LINKED ISSUE #202: Add specific scopes needed for API key
Currently the API key is defined with no scope enforcement. We could lock it down to only the specific scopes needed. I don't think these will change, at least not for this release so we can do this now. 

Currently these are set with [the setting:
](https://github.com/Harvard-University-iCommons/django-canvas-oauth#settings)
`CANVAS_OAUTH_SCOPES: []`

---------------------
To test:

- change your API Developer Key to enforce scopes and add the scopes listed in the default scopes file.

- Go to the admin, and delete your user's token record. Then relaunch and go through the OAuth process. Then look at the logs for an INFO level message from [oauth.py:73], where you should see the OAuth redirect to Canvas with the scopes in the URL, similar to the following message: 
` [INFO] [oauth.py:73] Redirecting user to https://canvas-test.it.umich.edu/login/oauth2/auth?client_id=17700000000000271&redirect_uri=https%3A%2F%2F4ac5-141-213-175-192.ngrok.io%2Foauth%2Foauth-callback&response_type=code&scope=url%3AGET%7C%2Fapi%2Fv1%2Fcourses%2F%3Acourse_id%2Ftabs+url%3APUT%7C%2Fapi%2Fv1%2Fcourses%2F%3Acourse_id%2Ftabs%2F%3Atab_id&state=7tHHBmfJyMKI`

- Remove one of the scopes in the API Developer Key, then add only the remaining scopes to the environment variable. Delete your user's token in the admin, then re-launch and authorize. The handshake should work, but the application will throw an error when it tries to use the scope you are missing:
`Error occurred! Status: 401; Message: An error occurred while communciating with Canvas. "Insufficient scopes on access token.`

### PR DESCRIPTION
**Pull request by ssciolla**
The PR aims to resolve #202.

### CODE REVIEW

File: backend/canvas_scopes.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+DEFAUlT_CANVAS_SCOPES = [
```
Comment (zqian) [inline backend/canvas_scopes.py:1]: maybe we should put those values as default setting in .env.sample file?

File: backend/settings.py [hunk 2]
```diff
@@ -221,7 +223,13 @@
 CANVAS_OAUTH_CLIENT_ID = os.getenv('CANVAS_OAUTH_CLIENT_ID', 'canvas_app_explorer')
 CANVAS_OAUTH_CLIENT_SECRET = os.getenv('CANVAS_OAUTH_CLIENT_SECRET', 'canvas_app_explorer')
 CANVAS_OAUTH_CANVAS_DOMAIN = os.getenv('CANVAS_OAUTH_CANVAS_DOMAIN', 'canvas.instructure.com')
-CANVAS_OAUTH_SCOPES = os.getenv('CANVAS_OAUTH_SCOPES', '').split(',')
+
+# Scopes environment variable provides a way to recover if Canvas changes scope identifiers.
```
Comment (zqian) [inline backend/settings.py:227]: How can I verify the OAuth access token is created with proper scope?

File: .env.sample [hunk 3]
```diff
@@ -15,6 +15,10 @@
 # (required) The domain of your canvas instance (e.g. canvas.instructure.com)
 # CANVAS_OAUTH_CANVAS_DOMAIN=
 
+# (optional) The Canvas API scopes needed by the application
```
Comment (zqian) [inline .env.sample:18]: the CANVAS_OAUTH_SCOPES is "A list of Canvas API scopes that the access token will provide access to.", as specified in the [django-canvas-oauth library](https://github.com/Harvard-University-iCommons/django-canvas-oauth/blob/master/canvas_oauth/settings.py). 

So the scopes are associated with user OAuth tokens, not the API key as described in the #202?

File: backend/canvas_scopes.py (hunk 1 above)
Comment (jonespm) [reply backend/canvas_scopes.py:1]: Yeah, I'd probably put these 3 in the config.

File: .env.sample (hunk 3 above)
Comment (ssciolla) [reply .env.sample:18]: Well, it's both I think. Scopes on the API Developer Key in Canvas and in the application configuration need to match (if you are enforcing scopes). When the user authorizes the application, the OAuth process checks whether the scopes match, and then mints an access token and refresh token pair which are associated with the scopes at the time they were created. The access token can only be used to make requests to those endpoints/scopes.

Comment (ssciolla) [reply .env.sample:18]: Like @jonespm said, I think it's possible the application requests fewer scopes than are enabled on the Canvas API Developer Key side, and it should still work. I'll test it to be sure.

File: backend/canvas_scopes.py (hunk 1 above)
Comment (ssciolla) [reply backend/canvas_scopes.py:1]: Done. https://github.com/tl-its-umich-edu/canvas-app-explorer/pull/245/commits/9be2ef56d714325811b73dd16f24fe97fda9c68b

File: .env.sample (hunk 3 above)
Comment (ssciolla) [reply .env.sample:18]: Yeah, it seems to work if you add a new scope that you don't end up using; OAuth process and requests work. I assume if you tried to use that missing scope using Postman or some application process, you'd still get an "insufficient scopes on token" error.

File: backend/canvas_scopes.py (hunk 1 above)
Comment (zqian) [reply backend/canvas_scopes.py:1]: Verified.

File: backend/settings.py (hunk 2 above)
Comment (zqian) [reply backend/settings.py:227]: Great! I followed those steps and verified.

1. log message as `[INFO] [oauth.py:73] Redirecting user to https://canvas-test.it.umich.edu/login/oauth2/auth?client_id=17700000000000271&redirect_uri=https%3A%2F%2F4ac5-141-213-175-192.ngrok.io%2Foauth%2Foauth-callback&response_type=code&scope=url%3AGET%7C%2Fapi%2Fv1%2Fcourses%2F%3Acourse_id%2Ftabs+url%3APUT%7C%2Fapi%2Fv1%2Fcourses%2F%3Acourse_id%2Ftabs%2F%3Atab_id&state=7tHHBmfJyMKI`
2. I removed the url:GET|/api/v1/courses/:id, removed the user oauth token in CAE admin page, and re-authenticated myself to CAE. Now, I got this error message "Error occurred! Status: 401; Message: An error occurred while communciating with Canvas. "Insufficient scopes on access token."

I will add those test steps to the original issue, and close this comment.

File: .env.sample (hunk 3 above)
Comment (zqian) [reply .env.sample:18]: Got it!

**Review by zqian — APPROVED** [review]

File: backend/settings.py (hunk 2 above)
Comment (ssciolla) [reply backend/settings.py:227]: I added a couple more test cases, @zqian, to the comment above.

**Review by jonespm — APPROVED** [review]

**Pull request merged**